atc/scheduler/factory/build_factory.go

```diff
@@ -305,6 +305,11 @@ func (factory *buildFactory) applyHooks(cp constructionParams) (atc.Plan, error)
 		return atc.Plan{}, err
 	}
 
+	cp, err = factory.errorIfPresent(cp)
```

Comment:
FWIW, I think this is where precedence becomes important. If you want the `on_error` hook to only act in response to the original step, and none of its hooks, I think putting it first here makes sense. Right now it's after the `on_abort` hook but it looks like that's fine because the `on_abort` hook [drops any errors from its hook on the floor](https://github.com/concourse/concourse/blob/master/atc/exec/on_abort.go#L36).

On the other hand, if you want the hook to fire if anything errors, including success/failure hooks on the same step, you'd want to put it last. But I think having it respond only to the original step *probably* makes the most sense - if you have an `on_failure` hook for cleanup and an `on_error` hook for cleanup, it probably wouldn't help to have one cause the other.

Reply:
In the case of `on_abort` dropping it errors, do we want to move this up higher in case that error ever gets handled? It is working for now but could become problematic at a later date.

Reply:
Hmm maybe it's fine as it is. Either way there's kind of a danger of one breaking the other, since they both check the `err` return value right now, so we'll just need to be careful and trust our tests.